doc/compile-scheme.1
.TH compile-scheme 1
.SH NAME
compile-scheme \- Compiling Scheme programs to executables
.SH SYNOPSIS
.P
\fBcompile-scheme\fR [\fB-A\fR \fIpath\fR] [\fB-I\fR \fIpath\fR] [\fB-o\fR \fIpath\fR] \fIinput-file.sps\fR
\#\fBcompile-scheme\fR [\fB-A\fR \fIpath\fR] [\fB-I\fR \fIpath\fR] [\fB-o\fR \fIpath\fR] [\fB-t\fR \fItarget\fR] \fIinput-file.sps\fR
.P
\fBcompile-scheme\fR [\fB-A\fR \fIpath\fR] [\fB-I\fR \fIpath\fR] [\fB-o\fR \fIpath\fR] \fIinput-file.scm\fR
\#\fBcompile-scheme\fR [\fB-A\fR \fIpath\fR] [\fB-I\fR \fIpath\fR] [\fB-o\fR \fIpath\fR] [\fB-t\fR \fItarget\fR] \fIinput-file.scm\fR
.P
\fBcompile-scheme\fR \fIlong-option\fR
.SH DESCRIPTION
.P
\fBcompile-scheme\fR is a tool to compile R6RS and R7RS Scheme programs.
It supports most of SRFI-138 but also adds more features.
.P
The program \fIinput-file\fR is compiled into an executable file.
The resulting executable file is written to file specified by the \fB-o\fR \fIpath\fR (if present) or to the file named same as \fIinput-file\fR but without the .scm or .sps suffix.
On Windows either .bat or .exe is appended to the output name.
.SH SUPPORTED IMPLEMENTATIONS
.P
Some implementations support both compiling and interpreting, in that
case only the compiler functionality is used and the implementation is marked
as compiler.
.SS "R6RS Compilers"
loko
.SS "R6RS Interpreters"
chezscheme guile ikarus ironscheme mosh racket sagittarius ypsilon
.SS "R7RS Compilers"
chicken cyclone loko
.SS "R7RS Interpreters"
chibi foment gauche guile kawa larceny meevax mit-scheme mosh racket sagittarius skint stklos tr7 ypsilon
.SH OPTIONS
.P
\fB\-A\fR \fIpath\fR
Append path to the list of directories that are searched in order to locate imported libraries.
.P
\fB\-I\fR \fIpath\fR
Prepend directory to the list of directories that are searched in order to locate imported libraries.
.P
\fB-o\fR \fIoutput-file\fR
Use the \fIoutput-file\fR file, instead of the default deducted from input-file, for the executable file produced.
.P
Multiple instances of the -A, and -I options can be specified.
\#.P
\#\fB\-t\fR { \fIunix\fR | \fIwindows\fR | \fIphp\fR }
\#\fB\-t\fR \fIunix\fR
\#Set the compilation target.
\#This is not needed if you are compiling for the target OS you are running.
\#Cross compilation is only supported in following cases:
\#.IP
\#From \fIunix\fR host to \fIwindows\fR target when chosen implementation is interpreter.
\#.IP
\#From \fIwindows\fR host to \fIunix\fR target when chosen implementation is interpreter.
\#.IP
\#From \fIunix\fR host to \fIphp\fR target when chosen implementation is interpreter.
\#.IP
\#From \fIwindows\fR host to \fIphp\fR target when chosen implementation is interpreter.
.P
.B \-\-list-r6rs
List supported R6RS implementations.
.P
.B \-\-list-r7rs
List supported R7RS implementations.
.P
.B \-\-list-all
List all supported implementations.
\#.P
\#\fB\-\-list-targets\fR
\#List all supported compilation targets.
.P
\fB\-\-version\fR
Show the software version.
\fB\-\-debug\fR
Turn on debug output.
.P
\fB\-\-help\fR
Shows you command to read this manual page. :)
.SH ENVIRONMENT
.P
\fBCOMPILE_R7RS\fR
.P
\fBCOMPILE_SCHEME\fR
.IP

Either of these environment variables \fBmust\fR be set.
Set either to the \fBname\fR of the implementation as specified in the support list.
\fBThis differs from SRFI-138\fR as the SRFI excepts a path.

.P
\fBCOMPILE_SCHEME_DEBUG\fR
.IP
Another way in addition to --debug to turn on debug output.
.SH STANDARDS
.P
SRFI 138: Compiling Scheme programs to executables.
https://srfi.schemers.org/srfi-138/srfi-138.html
.SH CAVEATS
.P
Differences from SRFI-138
.IP
No support for -D flag.
.IP
Not all implementations support adding to beginning or end of load path so -I and -A might work the same
.IP
Only supports one input-file.
.SH EXAMPLES
.P
Compile R6RS file with all dependencies in the same directory.
.IP
COMPILE_SCHEME=SCHEME compile-scheme main.sps
.P
Compile R7RS file with all dependencies in the same directory.
.IP
COMPILE_SCHEME=SCHEME compile-scheme main.scm
.P
Compile R6RS file with dependencies in libs directory.
.IP
COMPILE_SCHEME=SCHEME compile-scheme -I ./libs main.sps
.P
Compile R7RS file with dependencies in libs directory.
.IP
COMPILE_SCHEME=SCHEME compile-scheme -I ./libs main.scm
.P
Compile R6RS file with dependencies in libs directory, to output named foo.
.IP
COMPILE_SCHEME=SCHEME compile-scheme -I ./libs -o foo main.sps
.P
Compile R7RS file with dependencies in libs directory, to output named foo.
.IP
COMPILE_SCHEME=SCHEME compile-scheme -I ./libs -o foo main.scm
.P
Makefile for interpreter
.IP
.nf
.eo
PREFIX=/usr/local
SCHEME=chibi

build:
	COMPILE_SCHEME=${SCHEME} compile-scheme -I ${PREFIX}/lib/myapp/librarydirectory -o myapp myapp.scm

install:
	mkdir -p ${PREFIX}/lib/myapp
	cp -r librarydirectory ${PREFIX}/lib/myapp/
	cp  myapp.scm ${PREFIX}/lib/myapp/myapp.scm
	install myapp ${PREFIX}/bin/myapp
.ec
.fi



compile-scheme.scm
(import (scheme base)
        (scheme file)
        (scheme read)
        (scheme write)
        (scheme process-context)
        (libs util)
        (libs implementations)
        (libs library-util)
        (srfi 170))

(define debug?
  (if (or (member "--debug" (command-line))
          (get-environment-variable "SCHEME_COMPILE_DEBUG"))
    #t
    #f))

(when (member "--help" (command-line))
  (display "For help see: man compile-scheme")
  (newline)
  (exit 0))

(when (member "--version" (command-line))
  (display "1.1.1")
  (newline)
  (exit 0))

(when (or (member "--list-r6rs" (command-line))
        (member "--list-r6rs-schemes" (command-line)))
  (for-each (lambda (scheme) (display scheme) (display " ")) r6rs-schemes)
  (newline)
  (exit 0))

(when (member "--list-r6rs-except" (command-line))
  (for-each
    (lambda (scheme)
      (when (not (member (symbol->string scheme)
                         (cdr (member "--list-r6rs-except" (command-line)))))
        (display scheme)
        (display " ")))
    r6rs-schemes)
  (newline)
  (exit 0))

(when (or (member "--list-r7rs" (command-line))
          (member "--list-r7rs-schemes" (command-line)))
  (for-each (lambda (scheme) (display scheme) (display " ")) r7rs-schemes)
  (newline)
  (exit 0))

(when (member "--list-r7rs-except" (command-line))
  (for-each
    (lambda (scheme)
      (when (not (member (symbol->string scheme)
                         (cdr (member "--list-r7rs-except" (command-line)))))
        (display scheme)
        (display " ")))
    r7rs-schemes)
  (newline)
  (exit 0))

(when (or (member "--list-all" (command-line))
          (member "--list-schemes" (command-line)))
  (for-each (lambda (scheme) (display scheme) (display " ")) all-schemes)
  (newline)
  (exit 0))

(define scheme
  (cond
    ((get-environment-variable "COMPILE_R7RS")
     (string->symbol (get-environment-variable "COMPILE_R7RS")))
    ((get-environment-variable "COMPILE_SCHEME")
     (string->symbol (get-environment-variable "COMPILE_SCHEME")))
    (else #f)))

(when (not scheme)
  (display "Either environment variable COMPILE_R7RS or COMPILE_SCHEME is not set." (current-error-port))
  (newline (current-error-port))
  (exit 1))

(when (not (assoc scheme data)) (error "Unsupported implementation" scheme))

(define input-file
  (let ((input-file #f))
    (for-each
      (lambda (item)
        (if (or (string-ends-with? item ".scm")
                (string-ends-with? item ".sps"))
          (set! input-file item)))
      (list-tail (command-line) 1))
    (when (not input-file)
      (error "The intput-file must be either .scm (R7RS) or .sps (R6RS)"))
    input-file))

(define single-library-input-file
  (let ((input-file #f))
    (for-each
      (lambda (item)
        (when (or (string-ends-with? item ".sld")
                  (string-ends-with? item ".sls"))
          (set! input-file item)))
      (list-tail (command-line) 1))
    input-file))

(define r6rs?
  (if (and input-file
           (or (string-ends-with? input-file ".sps")
               (string-ends-with? input-file ".sls")))
    #t
    #f))

(define scheme-type (cdr (assoc 'type (cdr (assoc scheme data)))))

(define compilation-target
  (let ((outfile (if (member "-o" (command-line))
                   (cadr (member "-o" (command-line)))
                   (if input-file
                     "a.out"
                     #f))))
    (if (and (symbol=? scheme-type 'compiler))
      (string-append outfile ".bin")
      outfile)))

(define compilation-target
  (cond
    ((member "-t" (command-line))
     (cadr (member "-t" (command-line))))
    (else
      (cond-expand (windows 'windows)
                   (else 'unix)))))

(define output-file
  (let ((outfile
          (cond
            ((member "-o" (command-line))
             (cadr (member "-o" (command-line))))
            (input-file (string-cut-from-end input-file 4)))))
    outfile))

(define prepend-directories
  (letrec ((looper (lambda (rest result)
                     (if (null? rest)
                       result
                       (if (string=? (car rest) "-I")
                         (looper (cdr (cdr rest))
                                 (append (list (cadr rest)) result))
                         (looper (cdr rest)
                                 result))))))
    (looper (command-line) (list))))

(define append-directories
  (letrec ((looper (lambda (rest result)
                     (if (null? rest)
                       result
                       (if (string=? (car rest) "-A")
                         (looper (cdr (cdr rest))
                                 (append (list (cadr rest)) result))
                         (looper (cdr rest)
                                 result))))))
    (looper (command-line) (list))))

(when (member "--library-dependencies" (command-line))
  (write (library-dependencies scheme
                               (append prepend-directories append-directories)
                               (if input-file
                                 input-file
                                 single-library-input-file)
                               (list)
                               (list)))
  (newline)
  (exit 0))

(define library-files
  (library-dependencies scheme
                        (append prepend-directories append-directories)
                        (if input-file
                          input-file
                          single-library-input-file)
                        (list)
                        (list)))

(define scheme-command
  (apply (cdr (assoc 'command (cdr (assoc scheme data))))
         (list
           (cond
             ((symbol=? compilation-target 'windows) "")
             (else "exec"))
           (cond
             ((symbol=? compilation-target 'windows) "%0%")
             (else "\"$0\""))
           (cond
             ((symbol=? compilation-target 'windows) "")
             (else "\"$@\""))
           (if input-file input-file "")
           (if output-file output-file "")
           prepend-directories
           append-directories
           library-files
           r6rs?
           compilation-target)))

(define scheme-library-command
  (lambda (library-file)
    (apply (cdr (assoc 'library-command (cdr (assoc scheme data))))
      (list library-file prepend-directories append-directories r6rs?))))

(define list-of-features
  (letrec ((looper (lambda (rest result)
                     (if (null? rest)
                       result
                       (if (string=? (car rest) "-D")
                         (looper (cdr (cdr rest))
                                 (append (list (cadr rest)) result))
                         (looper (cdr rest)
                                 result))))))
    (looper (command-line) (list))))

(when (not (null? library-files))
  (when (assoc 'library-command (cdr (assoc scheme data)))
    (for-each
      (lambda (file)
        (let* ((library-command (scheme-library-command file)))
          (when debug?
            (display "[DEBUG] library-command: ")
            (write library-command)
            (newline))
          (for-each
            (lambda (command)
              (let ((exit-code (system command)))
                (when (not (= exit-code 0))
                  (exit exit-code))))
            library-command)))
      library-files)))

(when debug?
  (display "[DEBUG] scheme-command: ")
  (write scheme-command)
  (newline))

(when (and (equal? scheme-type 'interpreter) input-file)
  (when (and output-file (file-exists? output-file))
    (delete-file output-file))
  (let ((scheme-program (slurp input-file)))
    (with-output-to-file
      output-file
      (lambda ()
        (cond
          ((symbol=? compilation-target 'windows)
           (for-each
             display
             `(";dir; start /WAIT " ,scheme-command " && exit"
               #\newline
               ,scheme-program
               )))
          (else
            (for-each
              display
              `(#\newline
                "#|"
                #\newline
                ,scheme-command
                #\newline
                "|#"
                #\newline
                ,scheme-program))))))
    (cond ((symbol=? compilation-target 'unix)
           (system (string-append "chmod +x " output-file))))))

(when (and (symbol=? scheme-type 'compiler) input-file)
  (when (and output-file (file-exists? output-file))
    (delete-file output-file))
  (for-each
    (lambda (command)
      (let ((exit-code (system command)))
        (when (not (= exit-code 0))
          (exit exit-code))))
    scheme-command))
